miden-lib/asm/miden/kernels/tx/tx.masm

```diff
@@ -64,29 +64,33 @@ end
 
 #! Creates a new note and returns a pointer to the memory address at which the note is stored.
 #!
-#! Inputs: [ASSET, tag, RECIPIENT]
+#! Inputs: [note_type, ASSET, tag, RECIPIENT]
 #! Outputs: [ptr, 0, 0, 0, 0, 0, 0, 0, 0]
 #!
 #! ASSET is the asset to be included in the note.
 #! tag is the tag to be included in the note.
 #! RECIPIENT is the recipient of the note.
 #! ptr is the pointer to the memory address at which the note is stored.
 export.create_note
+    # Move the note type down the stack
+    movdn.6
+    # => [ASSET, tag, note_type, RECIPIENT]
```

Comment:
Since the first thing we do here is re-arrange the stack, would it make sense to just define the inputs to the procedure to be `[ASSET, tag, note_type, RECIPIENT]`?

Also, we should probably check somewhere here that the `note_type` is valid.

Reply:
> Also, we should probably check somewhere here that the note_type is valid.

AFAIU this is the library code, which can not be called directly from user code. That is to say, this data is not user provided, but provided by the kernel shims, and those always push correct constants.

> Since the first thing we do here is re-arrange the stack, would it make sense to just define the inputs to the procedure to be [ASSET, tag, note_type, RECIPIENT]?

AFAICT this would not improve features/readability and would duplicate code in the `create_offchain_note`, `create_public_note`, and `create_encrypted_note`. I don't think we need this change

Reply:
> AFAIU this is the library code, which can not be called directly from user code. That is to say, this data is not user provided, but provided by the kernel shims, and those always push correct constants.

Yep - make sense.

> AFAICT this would not improve features/readability and would duplicate code in the `create_offchain_note`, `create_public_note`, and `create_encrypted_note`. I don't think we need this change

Why would this duplicate the code? I was thinking it might make the code a bit simpler - but maybe I'm missing something?

Reply:
> Why would this duplicate the code? I was thinking it might make the code a bit simpler - but maybe I'm missing something?

It duplicates the `movdn.6` from here, to the `create_offchain_note` and `create_public_note`. Not that big of a deal, changed.

Reply:
Thank you! There is a bigger decision to be made about how these create methods should work - especially in the context of on-chain notes ([see here](https://github.com/0xPolygonMiden/miden-base/issues/402#issuecomment-1995652092)).